keylime/signing.py

```diff
@@ -83,3 +92,49 @@
         verified = False
 
     return verified
+
+
+def verify_dsse_envelope(envelope: bytes, key: Optional[bytes] = None):
+    if key:
+        verifiers = [("user_provided_key", dsse_ecdsa.Verifier(load_pem_public_key(key)))]
+    else:
+        envelope_json = json.loads(envelope)
+
+        certificates = [base64.b64decode(signer["keyid"]) for signer in envelope_json["signatures"]]
+        verifiers = [
+            (hashlib.sha256(cert), dsse_x509.Verifier(load_pem_x509_certificate(cert))) for cert in certificates
+        ]
+
+    verification_result = None
+    try:
+        verification_result = dsse.Verify(envelope.decode(), verifiers)
+    except ValueError as e:
+        logger.warning("Unable to verify DSSE envelope: %s", e)
+
+    return verification_result
+
+
+def get_runtime_policy_keys(runtime_policy: bytes, pubkey: Optional[str] = None):
+    runtime_policy_json = json.loads(runtime_policy)
+    if runtime_policy_json.get("payload"):
+        pubkey_dir = "/var/lib/keylime/signing_keys/"
+        if pubkey:
+            runtime_policy_key_bytes = base64.b64decode(pubkey)
+            keyid = hashlib.sha256(runtime_policy_key_bytes).hexdigest()
+            pubkey_path = pubkey_dir + keyid + ".pub"
+            logger.info("Writing provided DSSE public key to disk at %s", pubkey_path)
+            if not os.path.exists(pubkey_dir):
+                os.makedirs(pubkey_dir)
+            with open(pubkey_path, "wb") as f:
+                f.write(runtime_policy_key_bytes)
+        else:
+            runtime_policy_key_bytes = None
+            for sig in runtime_policy_json["signatures"]:
+                keyid = sig["keyid"]
+                pubkey_path = pubkey_dir + keyid + ".pub"
+                if os.path.isfile(pubkey_path):
```

Comment:
@mbestavros I assume these security warnings are fine to dismiss because we need to use the user input for the filename and we aren't just letting them reference a full file path or doing things like include wildcards or "../" by forcing the prefix and the suffix. But just wanted to confirm.

Reply:
@mpeters I'd say so, especially because it isn't even "user-provided" per se - this code writes keyfiles to disk, and it just names them with the SHA256 hash of the keyfile (to find it later). So the only "arbitrary input" here is the hash of the file. If an adversary is able to hijack SHA256, we have much bigger problems :)